packages/application-shell/src/components/navbar/use-navbar-state-manager.ts

```diff
@@ -0,0 +1,245 @@
+        isMenuOpen: false,
+      };
+    default:
+      return state;
+  }
+};
+
+const useNavbarStateManager = (props: HookProps) => {
+  const navBarNode = React.useRef<HTMLElement>(null);
+  const areProjectExtensionsEnabled = useFeatureToggle(PROJECT_EXTENSIONS);
+  const applicationsNavBarMenu = useApplicationsMenu<'navBar'>('navBar', {
+    queryOptions: {
+      onError: reportErrorToSentry,
+    },
+    skipRemoteQuery: !props.environment.servedByProxy,
+    loadMenuConfig: props.DEV_ONLY__loadNavbarMenuConfig,
+  });
+  const { data: projectExtensionsQuery } = useQuery<
+    TFetchProjectExtensionsNavbarQuery,
+    TFetchProjectExtensionsNavbarQueryVariables
+  >(FetchProjectExtensionsNavbar, {
+    skip: !props.environment.servedByProxy || !areProjectExtensionsEnabled,
+    variables: {
+      target: GRAPHQL_TARGETS.SETTINGS_SERVICE,
+    },
+    fetchPolicy: 'cache-and-network',
+    onError: reportErrorToSentry,
+  });
+  const customAppsMenu =
+    projectExtensionsQuery &&
+    projectExtensionsQuery.projectExtension &&
+    projectExtensionsQuery.projectExtension.applications
+      ? projectExtensionsQuery.projectExtension.applications
+          .map<TNavbarMenu | undefined>(app => {
+            // Map the menu properties to match the one from the proxy schema.
+            // This is to ensure that the menu object is the same from the proxy
+            // config and the custom apps config, thus allowing them to be
+            // concatenated and rendered the same way.
+            if (!app.navbarMenu) return;
+            return {
+              key: app.navbarMenu.key,
+              uriPath: app.navbarMenu.uriPath,
+              labelAllLocales: app.navbarMenu.labelAllLocales || [],
+              icon: app.navbarMenu.icon,
+              featureToggle: app.navbarMenu.featureToggle,
+              permissions: app.navbarMenu.permissions as string[],
+              submenu: (app.navbarMenu.submenu || []).map(menu => ({
+                key: menu.key,
+                uriPath: menu.uriPath,
+                labelAllLocales: menu.labelAllLocales || [],
+                featureToggle: menu.featureToggle,
+                permissions: menu.permissions as string[],
+                menuVisibility: undefined,
+                actionRights: undefined,
+                dataFences: undefined,
+              })),
+              menuVisibility: undefined,
+              actionRights: undefined,
+              dataFences: undefined,
+              shouldRenderDivider: false,
+            };
+          })
+          .filter(nonNullable)
+      : [];
+  const cachedIsForcedMenuOpen = window.localStorage.getItem(
+    STORAGE_KEYS.IS_FORCED_MENU_OPEN
+  );
+  const isForcedMenuOpen = cachedIsForcedMenuOpen === 'true' || null;
+
+  const [state, dispatch] = React.useReducer<
+    (prevState: State, action: Action) => State
+  >(reducer, initialState);
+
+  const checkSize = React.useCallback(
+    throttle(() => {
```

Comment:
This is kind of like having a useThrottle, right? I wonder if having that would make more sense?